### ISSUE TITLE
Microsoft.Azure.Services.AppAuthentication AzureServicesAuthConnectionString Not Working on Guest OS 6 (Windows Server 2019 Datacenter)

### ISSUE THREAD

**Issue by wjchristenson2** [Azure/azure-sdk-for-net#5555]
We just changed our target guest OS family to 6 (Windows Server 2019 Datacenter edition). Our AzureServicesAuthConnectionString for Microsoft.Azure.Services.AppAuthentication (1.2.0-preview2) is no longer working. Certificate is present/good, the AzureServicesAuthConnectionString environment variable is present/good. Nothing has changed code-wise. 

Any reason why Microsoft.Azure.Services.AppAuthentication (1.2.0-preview2) will not work with Windows Server 2019 Datacenter edition? 2019 does come with .NET 4.7.2 - any hard dependencies on older versions throwing things off?

Attached = stacktrace.
[stacktrace.txt](https://github.com/Azure/azure-sdk-for-net/files/2998085/stacktrace.txt)

**Comment by kurtzeborn** [comment]
Thank you for opening this issue! We are routing it to the appropriate team for follow up.

**Comment by varunsh-msft** [comment]
As per the error message in the stack trace, no connection string is specified. This means the library did not find the environment variable AzureServicesAuthConnectionString. Can you please check that the env variable exists? If possible, try to read the print the env variable using your own code.

**Comment by varunsh-msft** [comment]
@wjchristenson2 , do you have an update on this? Thanks!

**Comment by wjchristenson2** [comment]
@varunsh-msft Update - I created a quick test controller action/view to display the environment variable and it worked great using the following code.

```
string azureServicesAuthConnectionString = System.Environment.GetEnvironmentVariable("AzureServicesAuthConnectionString", EnvironmentVariableTarget.Machine);

ViewBag.AzureServicesAuthConnectionString = azureServicesAuthConnectionString ?? "";

```
Returns the following connection string with no issue:

> RunAs=App;AppId=90bdfb3b-4f5c-4d7c-9542-c7f46fdb3c57;TenantId=c191f8c9-b4cb-47de-971b-dc01ee887392;CertificateThumbprint=37D386EA2CA0CC30A5E2E09DF487AE1AA4D8C7C3;CertificateStoreLocation=LocalMachine;

Even so, we are still getting the following error using Microsoft.Azure.Services.AppAuthentication (1.2.0-preview2) running on Windows Server 2019 Datacenter Edition (azure guest os family 6). When we revert back to azure guest os family 5, works fine.

> Exception: Microsoft.Azure.Services.AppAuthentication.AzureServiceTokenProviderException: Parameters: Connection String: [No connection string specified], Resource: https://vault.azure.net, Authority: https://login.windows.net/c191f8c9-b4cb-47de-971b-dc01ee887392. Exception Message: Tried the following 4 methods to get an access token, but none of them worked.
Parameters: Connection String: [No connection string specified], Resource: https://vault.azure.net, Authority: https://login.windows.net/c191f8c9-b4cb-47de-971b-dc01ee887392. Exception Message: Tried to get token using Managed Service Identity. Access token could not be acquired. MSI ResponseCode: BadRequest, Response: {"error":"invalid_request","error_description":"Identity not found"}
Parameters: Connection String: [No connection string specified], Resource: https://vault.azure.net, Authority: https://login.windows.net/c191f8c9-b4cb-47de-971b-dc01ee887392. Exception Message: Tried to get token using Visual Studio. Access token could not be acquired. Visual Studio Token provider file not found at "D:\Windows\system32\config\systemprofile\AppData\Local\.IdentityService\AzureServiceAuth\tokenprovider.json"
Parameters: Connection String: [No connection string specified], Resource: https://vault.azure.net, Authority: https://login.windows.net/c191f8c9-b4cb-47de-971b-dc01ee887392. Exception Message: Tried to get token using Azure CLI. Access token could not be acquired. 
Parameters: Connection String: [No connection string specified], Resource: https://vault.azure.net, Authority: https://login.windows.net/c191f8c9-b4cb-47de-971b-dc01ee887392. Exception Message: Tried to get token using Active Directory Integrated Authentication. Access token could not be acquired. get_user_name_failed: Failed to get user nameInner Exception : No mapping between account names and security IDs was done...

**Comment by varunsh-msft** [comment]
This is interesting. The library gets the env variable using this code:
https://github.com/Azure/azure-sdk-for-net/blob/c6dacb380c28bab71edc832ce91c66f742cf6fb4/src/SdkCommon/AppAuthentication/Azure.Services.AppAuthentication/AzureServiceTokenProvider.cs#L82

                connectionString = Environment.GetEnvironmentVariable("AzureServicesAuthConnectionString");

I see this is not using the EnvironmentVariableTarget.Machine argument. Will investigate further and get back.

**Comment by wjchristenson2** [comment]
@varunsh-msft It looks like the default w/o target pulls env variables from the current process (copied forward via parent process if its a child). So us adding them via a deployment startup task wouldn't be present unless we recycled the process. To test this theory, I tried recycling some processes to see if a new process would pull in the windows env variables from registry. I tried the following:

1. recycled the world wide web publishing service: didn't work
2. recycled the app pool: didn't work
3. recycled iis via its ui (believe same as 1): didn't work
4. rebooted the VM: worked!

With azure guest os 5 we had to install .net 4.7.2 via a startup task. This likely caused the server to reboot or recycle key processes that reloaded the environment variables. Now that azure guest os 6 has 4.7.2, we are not installing it which means those processes are not recycled on deploy so our environment variable we are adding via a startup command is not present.

I would think that the library should require a target of machine *or* have a a pecking order of target = machine then fallback to current process env variables. Something of the sort.

**Comment by varunsh-msft** [comment]
Ok, we should be able to fix this soon. Thanks!

**Comment by nonik0** [comment]
Hi @wjchristenson2, sorry but I missed updating this issue when we released [1.2.0-preview](https://www.nuget.org/packages/Microsoft.Azure.Services.AppAuthentication/1.2.0-preview3) a couple weeks ago. That release should fix this issue, so please give it a try and let us know if you have any further issues!

**Comment by wjchristenson2** [comment]
@nonik0 Just tried 1.2.0-preview3 and it seems to be working now. Thanks for the fix.
